compiler/mir/mirgen.nim

```diff
@@ -434,43 +385,175 @@ func nameNode(s: PSym): MirNode =
   else:
     unreachable(s.kind)
 
-template genLocation(c: var TCtx, n: PNode): EValue =
-  c.stmts.nodes.emit(nameNode(n.sym))
-
-func emit(dest: var MirBuffer, sp: var SourceProvider, src: MirBuffer,
-          span: NodeSpan): EValue =
-  ## Copies the nodes denoted by `span` from `src` to the end of `dest`
-  assert span.len > 0, "empty source fragment"
+func genLocation(c: var TCtx, n: PNode): Value =
+  let f = c.builder.push: c.builder.add(nameNode(n.sym))
+  c.builder.popSingle(f)
 
-  # make sure that the source information in `dest` is up-to-date before
-  # appending
-  applySource(dest, sp)
-
-  dest.nodes.add toOpenArray(src.nodes, span)
-
-  result = EValue(typ: dest.nodes[^1].typ)
-
-func genDefault(c: var TCtx, typ: PType): EValue =
-  ## Emits a call to the ``default`` operator for the given `typ`
-  argBlock(c.stmts): discard
-  magicCall(c.stmts.nodes, mDefault, typ)
+template allocTemp(c: var TCtx, typ: PType; alias=false): Value =
+  ## Allocates a new ID for a temporary and returns the name.
+  c.builder.allocTemp(typ, alias)
 
 proc gen(c: var TCtx; n: PNode)
-proc genx(c: var TCtx; n: PNode, consume: bool = false): EValue
+proc genx(c: var TCtx; n: PNode, consume: bool = false)
 proc genComplexExpr(c: var TCtx, n: PNode, dest: Destination)
 
 proc genAsgn(c: var TCtx, dest: Destination, rhs: PNode)
 proc genWithDest(c: var TCtx; n: PNode; dest: Destination)
 
-template selectArg(useConsume: bool): Sink {.dirty.} =
-  if useConsume: consume()
-  else:          arg()
+func getTemp(c: var TCtx, typ: PType): Value =
+  ## Allocates a new temporary and emits a definition for it into the
+  ## final buffer.
+  assert typ != nil
+  result = c.allocTemp(typ)
+  withFront c.builder:
+    c.subTree mnkDef:
+      c.use result
+      c.add MirNode(kind: mnkNone)
+
+template buildStmt(c: var TCtx, k: MirNodeKind, body: untyped) =
+  c.builder.buildStmt(k, body)
+
+template buildMagicCall(c: var TCtx, m: TMagic, t: PType, body: untyped) =
+  c.subTree MirNode(kind: mnkMagic, magic: m, typ: t):
+    body
+
+func detectKind(tree: MirTree, n: NodePosition, sink: bool): ExprKind =
+  ## Detects the kind of expression `n` (with the originating from AST `e`)
+  ## represents. `sink` informs whether expression is used in a sink context.
```

Comment:
Can you elaborate a bit on the logic behind the `sink` boolean? In particular, why is `sink` not a modifier applied to the node (ie. wrapping `sink(node)`) but must be derived by the caller?

My personal hunch is that this could be fragile due to information loss in new code paths (ie. the writer didn't take this into account), but I could be very wrong here.

Reply:
Hmm, I'm not sure I understand the questions. An expression itself doesn't know how or where it's used, only the usage site does.

If you mean that there should be an `nkSink` (or similar) node for annotating expression as appearing in a context where the should be sunken (i.e., ideally moved, copied if not possible), then I strongly agree -- this information is something that should be inferred during semantic analysis, not by the AST-to-MIR translation.

Reply:
> If you mean that there should be an `nkSink` (or similar) node for annotating expression as appearing in a context where the should be sunken (i.e., ideally moved, copied if not possible), then I strongly agree -- this information is something that should be inferred during semantic analysis, not by the AST-to-MIR translation.

Yep, that's what I meant. I take it as this is a planned change?

Reply:
Yep, from my side, it is, although not as part of this PR.